File: programmers/C_language/택배상자꺼내기.c
#include <stdio.h>

int solution(int n, int w, int num) {
    // row, col 저장용 2차원 배열
    int position[101][2];

    // 배열 생성
    for (int i = 0; i < n; i++) {
        int row = i / w;
        int temp = i % w;

        // 짝수 줄: 왼->오, 홀수 줄: 오->왼
        int col = (row % 2 == 0) ? temp : (w - 1 - temp);

        position[i + 1][0] = row;
        position[i + 1][1] = col;
    }

    // 찾는 값 저장
    int targetRow = position[num][0];
    int targetCol = position[num][1];

    int cnt = 0;

    for (int i = 1; i <= n; i++) {
        if (i == num)
            continue;

        int r = position[i][0];
        int c = position[i][1];

        // 같은 열이고 위에 있는 상자인지 확인
        if (c == targetCol && r > targetRow) {
            cnt++;
        }
    }

    return cnt + 1;
}

File: Java/level1/05_24/택배상자꺼내기.java
class Solution {
    public int solution(int n, int w, int num) {
        int answer = 0;

        // 필요한 행 개수
        int h = (n + w - 1) / w;

        // 상자 세팅
        int[][] boxes = new int[h][w];

        for (int i = 0; i < n; i++) {
            int level = i / w;   // 층
            int pos = i % w;     // 가로 위치

            // 짝수층: 왼 -> 오
            if (level % 2 == 0) {
                boxes[level][pos] = i + 1;
            }
            // 홀수층: 오 -> 왼
            else {
                boxes[level][w - 1 - pos] = i + 1;
            }
        }

        // 원하는 상자의 위치 찾기
        int targetRow = 0;
        int targetCol = 0;

        for (int i = 0; i < h; i++) {
            for (int j = 0; j < w; j++) {
                if (boxes[i][j] == num) {
                    targetRow = i;
                    targetCol = j;
                }
            }
        }

        // 같은 열에서 위에 있는 박스 개수 세기
        for (int i = targetRow; i < h; i++) {
            if (boxes[i][targetCol] != 0) {
                answer++;
            }
        }

        return answer;
    }
}